c_src/sqlite3_nif.c

```diff
@@ -632,6 +632,73 @@ exqlite_transaction_status(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
       autocommit == 0 ? make_atom(env, "transaction") : make_atom(env, "idle"));
 }
 
+static ERL_NIF_TERM
+exqlite_serialize(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
+{
+    assert(env);
+
+    connection_t* conn = NULL;
+    unsigned char* buffer = NULL;
+    sqlite3_int64 buffer_size = 0;
+    ERL_NIF_TERM serialized;
+
+    if (argc != 1) {
+        return enif_make_badarg(env);
+    }
+
+    if (!enif_get_resource(env, argv[0], connection_type, (void**)&conn)) {
+        return make_error_tuple(env, "invalid_connection");
+    }
+
+    buffer = sqlite3_serialize(conn->db, "main", &buffer_size, 0);
+    if (!buffer) {
+        return make_error_tuple(env, "serialization_failed");
+    }
+
+    serialized = make_binary(env, buffer, buffer_size);
+    sqlite3_free(buffer);
+
+    return make_ok_tuple(env, serialized);
+}
+
+static ERL_NIF_TERM
+exqlite_deserialize(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
+{
+    assert(env);
+
+    connection_t* conn = NULL;
+    unsigned char* buffer = NULL;
+    ErlNifBinary serialized;
+    int size = 0;
+    int rc = 0;
+    int flags = SQLITE_DESERIALIZE_FREEONCLOSE | SQLITE_DESERIALIZE_RESIZEABLE;
+
+    if (argc != 2) {
+        return enif_make_badarg(env);
+    }
+
+    if (!enif_get_resource(env, argv[0], connection_type, (void**)&conn)) {
+        return make_error_tuple(env, "invalid_connection");
+    }
+
+    if (!enif_inspect_binary(env, argv[1], &serialized)) {
+        return enif_make_badarg(env);
+    }
+
+    size = serialized.size;
+    buffer = sqlite3_malloc(size);
+    if (!buffer) {
+        return make_error_tuple(env, "deserialization_failed");
+    }
+
+    rc = sqlite3_deserialize(conn->db, "main", buffer, size, size, flags);
```

Comment:
The docs only mention `SQLITE_BUSY` as a potential error code, are there other errors I need to check for here?

Reply:
Checkout what we do in `exqlite_execute` for returning an error code.

```c
    rc = sqlite3_exec(conn->db, (char*)bin.data, NULL, NULL, NULL);
    if (rc != SQLITE_OK) {
        return make_sqlite3_error_tuple(env, rc, conn->db);
    }
```

This will cover any cases that change in the future.